In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [2]:
df = pd.read_csv('apartamentos.csv')

## Cleaning

First we drop the index column which was carried over from `collect_data_realestate.py` because we didn't use `index=False` as an argument of `to_csv`.

In [3]:
df.drop(columns=['Unnamed: 0'], inplace=True)

In [4]:
df

aluguel  \
0      \n\t\t\t\t\t\t\t\tR$ 4.700\n\t\t\t\t\t\t\t   
1              \n\t\t\t\t\t\tR$ 3.720\n\t\t\t\t\t   
2              \n\t\t\t\t\t\tR$ 3.300\n\t\t\t\t\t   
3              \n\t\t\t\t\t\tR$ 1.250\n\t\t\t\t\t   
4              \n\t\t\t\t\t\tR$ 2.835\n\t\t\t\t\t   
...                                           ...   
20731          \n\t\t\t\t\t\tR$ 2.100\n\t\t\t\t\t   
20732          \n\t\t\t\t\t\tR$ 1.500\n\t\t\t\t\t   
20733          \n\t\t\t\t\t\tR$ 2.700\n\t\t\t\t\t   
20734  \n\t\t\t\t\t\t\t\tR$ 1.000\n\t\t\t\t\t\t\t   
20735          \n\t\t\t\t\t\tR$ 1.000\n\t\t\t\t\t   

                                        condominio  \
0        \n\t\t+ R$ \n\t\t750 \n\t\tCondomínio\n\t   
1        \n\t\t+ R$ \n\t\t970 \n\t\tCondomínio\n\t   
2        \n\t\t+ R$ \n\t\t443 \n\t\tCondomínio\n\t   
3        \n\t\t+ R$ \n\t\t752 \n\t\tCondomínio\n\t   
4        \n\t\t+ R$ \n\t\t740 \n\t\tCondomínio\n\t   
...                                            ...   
20731  \n\t\t+ R$ \n\t\t1.000 \n\t\tCondomínio\n\t   
20732    \n\t\t+ R$ \n\t\t700 \n\t\tCondomínio\n\t   
20733    \n\t\t+ R$ \n\t\t560 \n\t\tCondomínio\n\t   
20734    \n\t\t+ R$ \n\t\t328 \n\t\tCondomínio\n\t   
20735  \n\t\t+ R$ \n\t\t1.100 \n\t\tCondomínio\n\t   

                              area                                 bairro  \
0      \n\t\t\t\t\t82 m²\n\t\t\t\t  \n Chácara Santo Antônio, São Paulo\n   
1      \n\t\t\t\t\t50 m²\n\t\t\t\t           \n Vila Olímpia, São Paulo\n   
2      \n\t\t\t\t\t85 m²\n\t\t\t\t        \n Vila das Mercês, São Paulo\n   
3      \n\t\t\t\t\t35 m²\n\t\t\t\t           \n Vila Mascote, São Paulo\n   
4      \n\t\t\t\t\t72 m²\n\t\t\t\t                  \n Saúde, São Paulo\n   
...                            ...                                    ...   
20731  \n\t\t\t\t\t55 m²\n\t\t\t\t           \n Vila Mariana, São Paulo\n   
20732  \n\t\t\t\t\t45 m²\n\t\t\t\t                  \n Moema, São Paulo\n   
20733  \n\t\t\t\t\t58 m²\n\t\t\t\t          \n Santa Cecília, São Paulo\n   
20734  \n\t\t\t\t\t36 m²\n\t\t\t\t             \n Bela Vista, São Paulo\n   
20735  \n\t\t\t\t\t45 m²\n\t\t\t\t                  \n Moema, São Paulo\n   

                                         quartos  \
0      \n\t\t\t\t\t\t\t\t2 dorm.\n\t\t\t\t\t\t\t   
1      \n\t\t\t\t\t\t\t\t1 dorm.\n\t\t\t\t\t\t\t   
2      \n\t\t\t\t\t\t\t\t2 dorm.\n\t\t\t\t\t\t\t   
3      \n\t\t\t\t\t\t\t\t1 dorm.\n\t\t\t\t\t\t\t   
4      \n\t\t\t\t\t\t\t\t3 dorm.\n\t\t\t\t\t\t\t   
...                                          ...   
20731  \n\t\t\t\t\t\t\t\t1 dorm.\n\t\t\t\t\t\t\t   
20732  \n\t\t\t\t\t\t\t\t1 dorm.\n\t\t\t\t\t\t\t   
20733  \n\t\t\t\t\t\t\t\t2 dorm.\n\t\t\t\t\t\t\t   
20734  \n\t\t\t\t\t\t\t\t1 dorm.\n\t\t\t\t\t\t\t   
20735  \n\t\t\t\t\t\t\t\t1 dorm.\n\t\t\t\t\t\t\t   

                                               banheiros  \
0      \n\t\t\t\t\t2\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
1      \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
2      \n\t\t\t\t\t2\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
3      \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
4      \n\t\t\t\t\t2\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
...                                                  ...   
20731  \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
20732  \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
20733  \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
20734  \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t...   
20735                                                NaN   

                                                 garagem  \
0      \n\t\t\t\t\t2\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n...   
1                                                    NaN   
2      \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tvaga\n\...   
3      \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tvaga\n\...   
4      \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tvaga\n\...   
...                                                  ...   
20731  \n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tvaga\n\...   
20732

#### aluguel
Let's begin cleaning the `aluguel` column. We want it to be a numeric column. At first sight it looks like it contains prices in the form `number` `dot` `number`. Let's see:

In [5]:
pattern_aluguel = r"([0-9]+\.[0-9]+)"
alu = df['aluguel'].str.contains(pattern_aluguel)

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/strings.py:1843: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


In [6]:
alu.value_counts()

True     20213
False      523
Name: aluguel, dtype: int64

So, actually, `523` rows do not contain that format. What are they, then?

In [7]:
df['aluguel'][~alu].unique()

array(['\n\t\t\t\t\t\tR$ 900\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 740\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 990\n\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\tR$ 998\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\tR$ 940\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 700\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 950\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 850\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 987\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 790\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 899\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 800\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 960\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 890\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 980\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 860\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 967\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 620\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 974\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 690\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 979\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 865\n\t\t\t\t\t',
       '\n\t\t\t\t\t\tR$ 750\n\t\t\t\t

Ok, so they are prices below 1000. So the decimal point might or might not be in the pattern we want.

In [8]:
pattern_aluguel2 = r"([0-9]+?\.?[0-9]+)"
alu2 = df['aluguel'].str.contains(pattern_aluguel2)
alu2.value_counts()

True    20736
Name: aluguel, dtype: int64

We caught all possible values of the rent, then. We used capture groups above so that we can use pattern_aluguel2 to extract the string we want:

In [9]:
df['aluguel'] = df['aluguel'].str.extract(pattern_aluguel2)

In [10]:
df['aluguel'] = df['aluguel'].str.replace('.','').astype(float)

In [11]:
df['aluguel']

0        4700.0
1        3720.0
2        3300.0
3        1250.0
4        2835.0
          ...  
20731    2100.0
20732    1500.0
20733    2700.0
20734    1000.0
20735    1000.0
Name: aluguel, Length: 20736, dtype: float64

#### condominio
Now we do the same thing for `condominio`

In [12]:
df['condominio'].unique()

array(['\n\t\t+ R$ \n\t\t750 \n\t\tCondomínio\n\t',
       '\n\t\t+ R$ \n\t\t970 \n\t\tCondomínio\n\t',
       '\n\t\t+ R$ \n\t\t443 \n\t\tCondomínio\n\t', ...,
       '\n\t\t+ R$ \n\t\t7.200 \n\t\tCondomínio\n\t',
       '\n\t\t+ R$ \n\t\t98 \n\t\tCondomínio\n\t',
       '\n\t\t+ R$ \n\t\t1.712 \n\t\tCondomínio\n\t'], dtype=object)

In [13]:
pattern_condominio = r"([0-9]?\.?[0-9]+)"
df['condominio'].str.contains(pattern_condominio).value_counts()

True    19355
Name: condominio, dtype: int64

In [14]:
df['condominio'] = df['condominio'].str.extract(pattern_condominio)
df['condominio'] = df['condominio'].str.replace('.','').astype(float)

In [15]:
df['condominio']

0         750.0
1         970.0
2         443.0
3         752.0
4         740.0
          ...  
20731    1000.0
20732     700.0
20733     560.0
20734     328.0
20735    1100.0
Name: condominio, Length: 20736, dtype: float64

#### area

In [16]:
df['area'].unique()

array(['\n\t\t\t\t\t82 m²\n\t\t\t\t', '\n\t\t\t\t\t50 m²\n\t\t\t\t',
       '\n\t\t\t\t\t85 m²\n\t\t\t\t', '\n\t\t\t\t\t35 m²\n\t\t\t\t',
       '\n\t\t\t\t\t72 m²\n\t\t\t\t', '\n\t\t\t\t\t45 m²\n\t\t\t\t',
       '\n\t\t\t\t\t67 m²\n\t\t\t\t', '\n\t\t\t\t\t100 m²\n\t\t\t\t',
       '\n\t\t\t\t\t40 m²\n\t\t\t\t', '\n\t\t\t\t\t57 m²\n\t\t\t\t',
       '\n\t\t\t\t\t70 m²\n\t\t\t\t', '\n\t\t\t\t\t118 m²\n\t\t\t\t',
       '\n\t\t\t\t\t55 m²\n\t\t\t\t', '\n\t\t\t\t\t53 m²\n\t\t\t\t',
       '\n\t\t\t\t\t62 m²\n\t\t\t\t', '\n\t\t\t\t\t84 m²\n\t\t\t\t',
       '\n\t\t\t\t\t25 m²\n\t\t\t\t', '\n\t\t\t\t\t200 m²\n\t\t\t\t',
       '\n\t\t\t\t\t71 m²\n\t\t\t\t', '\n\t\t\t\t\t43 m²\n\t\t\t\t',
       '\n\t\t\t\t\t61 m²\n\t\t\t\t', '\n\t\t\t\t\t33 m²\n\t\t\t\t',
       '\n\t\t\t\t\t60 m²\n\t\t\t\t', '\n\t\t\t\t\t127 m²\n\t\t\t\t',
       '\n\t\t\t\t\t90 m²\n\t\t\t\t', '\n\t\t\t\t\t15 m²\n\t\t\t\t',
       '\n\t\t\t\t\t80 m²\n\t\t\t\t', '\n\t\t\t\t\t138 m²\n\t\t\t\t',
       '\n\t\t\t\t\t485 m²\n\

In [17]:
pattern_area = r"([0-9]?\.?[0-9]+)"
df['area'].str.contains(pattern_area).value_counts()

True    20640
Name: area, dtype: int64

In [18]:
df['area'] = df['area'].str.extract(pattern_area)
df['area'] = df['area'].str.replace('.','').astype(float)
df['area']

0        82.0
1        50.0
2        85.0
3        35.0
4        72.0
         ... 
20731    55.0
20732    45.0
20733    58.0
20734    36.0
20735    45.0
Name: area, Length: 20736, dtype: float64

#### bairro

In [19]:
df['bairro'] = df['bairro'].str.split(',').str[0].str.strip()

In [20]:
df['bairro'].unique()

array(['Chácara Santo Antônio', 'Vila Olímpia', 'Vila das Mercês',
       'Vila Mascote', 'Saúde', 'Bela Vista', 'Vila Mariana',
       'Jardim São Savério', 'Casa Verde', 'Moema', 'Barra Funda',
       'Aclimação', 'Mooca', 'Vila Guarani(Zona Sul)', 'Campo Grande',
       'Vila Andrade', 'Vila Madalena', 'Pinheiros', 'Campo Belo',
       'Perdizes', 'Vila Romana', 'Jardim Aeroporto', 'Vila Matilde',
       'Jardim Paulista', 'Água Branca', 'Brooklin', 'Morumbi',
       'Santo Amaro', 'Vila Nova Conceição', 'Jardim Íris', 'Itaim Bibi',
       'Brás', 'Santana', 'Freguesia do Ó', 'Cerqueira César', 'Butantã',
       'Higienópolis', 'Casa Verde Alta', 'Vila Vera', 'Mandaqui',
       'Tatuapé', 'Jaguaré', 'Cangaíba', 'Jardim Ester Yolanda',
       'Ermelino Matarazzo', 'Santa Cecília', 'Vila São José (Ipiranga)',
       'Panamby', 'Cambuci', 'Parque da Mooca', 'Vila Pompéia',
       'Vila Leopoldina', 'Alto da Boa Vista', 'Jardins', 'Centro',
       'Consolação', 'Vila Regente Feijó', 'Po

#### quartos, banheiros e garagem

In [21]:
df['quartos'].unique()

array(['\n\t\t\t\t\t\t\t\t2 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t1 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t3 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t4 dorm.\n\t\t\t\t\t\t\t', nan,
       '\n\t\t\t\t\t\t\t\t6 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t5 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t7 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t10 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t9 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t145 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t12 dorm.\n\t\t\t\t\t\t\t',
       '\n\t\t\t\t\t\t\t\t11 dorm.\n\t\t\t\t\t\t\t'], dtype=object)

In [22]:
pattern = r"([0-9])"
df['quartos'].str.contains(pattern).value_counts()

True    20512
Name: quartos, dtype: int64

In [23]:
df['quartos'] = df['quartos'].str.extract(pattern)
df['quartos'] = df['quartos'].astype(float)
df['quartos']

0        2.0
1        1.0
2        2.0
3        1.0
4        3.0
        ... 
20731    1.0
20732    1.0
20733    2.0
20734    1.0
20735    1.0
Name: quartos, Length: 20736, dtype: float64

In [24]:
df['banheiros'].unique()

array(['\n\t\t\t\t\t2\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t3\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t5\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t4\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       nan,
       '\n\t\t\t\t\t6\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t7\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t8\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t9\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t11\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t10\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t6329\n\t\t\t\t\t\n\t\t\t\t\t\tban\n\t\t\t\t\t\n\t\t\t\t'],
      dtype=object)

In [25]:
df['banheiros'] = df['banheiros'].str.extract(pattern)
df['banheiros'] = df['banheiros'].astype(float)
df['banheiros']

0        2.0
1        1.0
2        2.0
3        1.0
4        2.0
        ... 
20731    1.0
20732    1.0
20733    1.0
20734    1.0
20735    NaN
Name: banheiros, Length: 20736, dtype: float64

In [26]:
df['garagem'].unique()

array(['\n\t\t\t\t\t2\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       nan,
       '\n\t\t\t\t\t1\n\t\t\t\t\t\n\t\t\t\t\t\tvaga\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t3\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t5\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t4\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t7\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t6\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t9\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t10\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t12\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t8\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t11\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\t\t\n\t\t\t\t',
       '\n\t\t\t\t\t14\n\t\t\t\t\t\n\t\t\t\t\t\tvagas\n\t\t\t\

In [27]:
df['garagem'] = df['garagem'].str.extract(pattern)
df['garagem'] = df['garagem'].astype(float)
df['garagem']

0        2.0
1        NaN
2        1.0
3        1.0
4        1.0
        ... 
20731    1.0
20732    1.0
20733    NaN
20734    NaN
20735    1.0
Name: garagem, Length: 20736, dtype: float64

#### anunciante

In [28]:
df['anunciante'].unique()

array(['/imobiliarias/quintoandar_15743101-imoveis.html',
       '/imobiliarias/strutura-imoveis_15769039-imoveis.html',
       '/imobiliarias/urbano-imoveis_47277001-imoveis.html',
       '/imobiliarias/igor-almeida-real-estate-agency_47031223-imoveis.html',
       '/imobiliarias/imoveis-pro_47370738-imoveis.html',
       '/imobiliarias/kms-associados_15515774-imoveis.html',
       '/imobiliarias/eden-imoveis_15440215-imoveis.html',
       '/imobiliarias/imoveis-varandas_15448780-imoveis.html',
       '/imobiliarias/azera-assessoria-imobiliaria-ltda_30017727-imoveis.html',
       '/imobiliarias/so-flats_47243704-imoveis.html',
       '/imobiliarias/century21-house-prime_15731858-imoveis.html', nan,
       '/imobiliarias/medeiros-negocios-imobiliarios-ltda-me_47008473-imoveis.html',
       '/imobiliarias/scheid-imoveis_47068660-imoveis.html',
       '/imobiliarias/tarcha-imoveis_15458431-imoveis.html',
       '/imobiliarias/broker-vip_47372424-imoveis.html',
       '/imobiliarias/nn-ne

In [29]:
pattern_anunciante = r"(?<=/imobiliarias/)(.+)(?=_)"

In [30]:
df['anunciante'].str.contains(pattern_anunciante).value_counts()

True    20242
Name: anunciante, dtype: int64

In [31]:
df['anunciante'] = df['anunciante'].str.extract(pattern_anunciante)
df['anunciante']

0                                   quintoandar
1                                   quintoandar
2                                   quintoandar
3                                   quintoandar
4                                   quintoandar
                          ...                  
20731                              arbo-imoveis
20732                                    kramer
20733                  c.t.sevi-venda-aluga-adm
20734    medeiros-negocios-imobiliarios-ltda-me
20735                              aquino-flats
Name: anunciante, Length: 20736, dtype: object

Now we have a clean dataframe:

In [32]:
df.head()

aluguel  condominio  area                 bairro  quartos  banheiros  \
0   4700.0       750.0  82.0  Chácara Santo Antônio      2.0        2.0   
1   3720.0       970.0  50.0           Vila Olímpia      1.0        1.0   
2   3300.0       443.0  85.0        Vila das Mercês      2.0        2.0   
3   1250.0       752.0  35.0           Vila Mascote      1.0        1.0   
4   2835.0       740.0  72.0                  Saúde      3.0        2.0   

   garagem   anunciante  
0      2.0  quintoandar  
1      NaN  quintoandar  
2      1.0  quintoandar  
3      1.0  quintoandar  
4      1.0  quintoandar

## Dealing with the `NaN`s

Now that we cleaned our data its time to deal with the missing values.

In [33]:
df.isnull().sum()

aluguel          0
condominio    1381
area            96
bairro           0
quartos        224
banheiros      463
garagem       4287
anunciante     494
dtype: int64

In [34]:
df['garagem'].unique()

array([ 2., nan,  1.,  3.,  5.,  4.,  7.,  6.,  9.,  8.])

We see the column with the most missing values is the garage column. Inspecting its possible values (i.e., the unique values) we see there's no zero in there; but we know some flats for renting don't have any garages.

We also see when there's no garage available the site leaves the spot empty. So the most reasonable Ansatz to fill the missing values here is to fill it with zeros.

In [35]:
df['garagem'].fillna(0, inplace=True)

In [36]:
df.isnull().sum()

aluguel          0
condominio    1381
area            96
bairro           0
quartos        224
banheiros      463
garagem          0
anunciante     494
dtype: int64

In [37]:
df[df['area'].isnull()]

aluguel  condominio  area                 bairro  quartos  banheiros  \
620     3930.0       930.0   NaN                  Moema      2.0        3.0   
1021    4300.0         NaN   NaN          Brooklin Novo      1.0        2.0   
1115    2200.0       420.0   NaN                Jaguaré      2.0        2.0   
1234    6000.0         NaN   NaN  Chácara Santo Antônio      1.0        1.0   
1532    1000.0       428.0   NaN          Vila Penteado      2.0        1.0   
...        ...         ...   ...                    ...      ...        ...   
19121   2006.0      1200.0   NaN               Brooklin      NaN        NaN   
19143   2006.0      1200.0   NaN               Brooklin      NaN        NaN   
19155   2146.0      1297.0   NaN           Vila Olímpia      NaN        NaN   
19836   2500.0         NaN   NaN          Jardim Textil      3.0        1.0   
19925   1800.0       838.0   NaN           Vila Olímpia      2.0        2.0   

       garagem                           anunciante  
620        1.0                               marcio  
1021       1.0  vitoria-flats-negocios-imobiliarios  
1115       1.0                mario-beltrao-imoveis  
1234       1.0                 leia-graca-ricchetti  
1532       1.0                     diamante-imoveis  
...        ...                                  ...  
19121      0.0                                housi  
19143      0.0                                housi  
19155      0.0                                housi  
19836      1.0                          ags-imoveis  
19925      0.0                               marcio  

[96 rows x 8 columns]

One of the key information when renting an apartament is the area it has. Missing area cannot be replaced by some Ansatz (like the mean for like apartaments), because we could have very large apartaments and very small ones in the same building, for example.

And since the missing values are only `.4%` of the total values (vide infra), we'll simply discard these missing rows

In [38]:
len(df[df['area'].isnull()])/len(df['area'])

0.004629629629629629

In [39]:
df.dropna(subset=['area'], inplace=True)

Now let's see the percentage of missing values in each column:

In [40]:
100*df.isnull().sum()/(df.count() + df.isnull().sum())

aluguel       0.000000
condominio    6.545543
area          0.000000
bairro        0.000000
quartos       0.959302
banheiros     2.097868
garagem       0.000000
anunciante    2.393411
dtype: float64

The most alarming missing value here is the condominio. The quartos and banheiros we can drop it -- they're not really expressive.

The anunciante is not really a very important variable, to begin with. So, for now, we'll just leave the missing values from anunciante as they are.

In [41]:
df.dropna(subset=['quartos', 'banheiros'], inplace=True)
100*df.isnull().sum()/(df.count() + df.isnull().sum())

aluguel       0.000000
condominio    6.369077
area          0.000000
bairro        0.000000
quartos       0.000000
banheiros     0.000000
garagem       0.000000
anunciante    2.269327
dtype: float64

Now, to deal with the missing data from condominio we have to note the following: the condominio is set by the building and its related to the services it offers.
Probably a good index for many services in the building is the garage: the more spots there's for a single apartament, probably the steeper the condominio is.

Indeed we see this relationship in the following correlation matrix

In [42]:
df.corr()

aluguel  condominio      area   quartos  banheiros   garagem
aluguel     1.000000    0.135154 -0.009415  0.223894   0.260540  0.213615
condominio  0.135154    1.000000  0.061756  0.502467   0.656740  0.674347
area       -0.009415    0.061756  1.000000  0.035033   0.038561  0.045013
quartos     0.223894    0.502467  0.035033  1.000000   0.744309  0.631318
banheiros   0.260540    0.656740  0.038561  0.744309   1.000000  0.741554
garagem     0.213615    0.674347  0.045013  0.631318   0.741554  1.000000

To fill the missing values from condominio we'll use the K nearest neighboors algorithm based on the garage:

In [43]:
df_nona = df.copy().dropna()
X_train, X_test, y_train, y_test = train_test_split(np.array(df_nona['garagem']).reshape(-1,1), df_nona['condominio'], random_state = 1)

In [44]:
knn = KNeighborsRegressor(n_neighbors=10)
knn.fit(X_train, y_train)
predictions_test = knn.predict(X_test)
predictions_train = knn.predict(X_train)

Next we'll compare the root mean squared error (RMSE) of the predctions from the train set and the test set to see if we're overfitting.

In [45]:
print(np.sqrt(mean_squared_error(y_test, predictions_test)))
print(np.sqrt(mean_squared_error(y_train, predictions_train)))
print('-'*100)
print(np.std(y_test))

720.7125136839882
727.1420014203217
----------------------------------------------------------------------------------------------------
965.3466892493418


So, we're really not overfitting -- if we were the RMSE from the train set would be much smaller than the one from the test set.

At the same time, the RMSE is smaller than the standard deviation, so its a reasonable estimate. We'll go ahead and use this to fill in the `NaN`s

In [46]:
df[df['condominio'].isnull()]['garagem']

32       0.0
58       1.0
74       0.0
76       1.0
93       1.0
        ... 
20605    1.0
20616    0.0
20632    0.0
20701    1.0
20717    1.0
Name: garagem, Length: 1277, dtype: float64

In [47]:
knn = KNeighborsRegressor(n_neighbors=10)
knn.fit(df_nona[['garagem']], df_nona['condominio'])

X_nan = np.array(df[df['condominio'].isnull()]['garagem']).reshape(-1,1)
predictions_nan = knn.predict(X_nan)

Now we'll substitute the missing values for the `predictions_nan`

In [48]:
df[df['condominio'].isnull()]

aluguel  condominio  area                    bairro  quartos  \
32      1430.0         NaN  15.0                  Perdizes      1.0   
58      4570.0         NaN  35.0                Bela Vista      1.0   
74      1400.0         NaN  35.0  Vila São José (Ipiranga)      1.0   
76      2400.0         NaN  65.0                   Morumbi      1.0   
93      3240.0         NaN  41.0                    Centro      1.0   
...        ...         ...   ...                       ...      ...   
20605   2900.0         NaN  42.0                   Jardins      1.0   
20616   1100.0         NaN  40.0                     Mooca      1.0   
20632   1100.0         NaN  40.0                     Mooca      1.0   
20701   2900.0         NaN  42.0                   Jardins      1.0   
20717   2900.0         NaN  42.0                   Jardins      1.0   

       banheiros  garagem                          anunciante  
32           1.0      0.0                         quintoandar  
58           1.0      1.0                         quintoandar  
74           1.0      0.0  icasaa_super-destaques-e-destaques  
76           1.0      1.0                            so-flats  
93           1.0      1.0                         quintoandar  
...          ...      ...                                 ...  
20605        1.0      1.0                            so-flats  
20616        1.0      0.0                         quintoandar  
20632        1.0      0.0                         quintoandar  
20701        1.0      1.0                            so-flats  
20717        1.0      1.0                            so-flats  

[1277 rows x 8 columns]

First we create a pandas series from `predictions_nan` with the same index as the null values from `condominio`:

In [49]:
values_nan = pd.Series(predictions_nan, index=df[df['condominio'].isnull()].index)
values_nan

32       393.3
58       969.1
74       393.3
76       969.1
93       969.1
         ...  
20605    969.1
20616    393.3
20632    393.3
20701    969.1
20717    969.1
Length: 1277, dtype: float64

And then we fill the missing values:

In [50]:
df['condominio'].fillna(values_nan, inplace=True)

In [51]:
df['condominio'].isnull().sum()

0

## Making sense of the data

Now, we'll explore the data itself and see if there's anything wrong with the values it contains.

In [52]:
df.describe()

aluguel    condominio           area       quartos     banheiros  \
count  20050.000000  20050.000000   20050.000000  20050.000000  20050.000000   
mean    2736.777007   1092.741436     143.111671      1.946633      1.813317   
std     1819.667976    973.801082    3505.853559      0.952430      1.230529   
min       10.000000      1.000000       1.000000      1.000000      1.000000   
25%     1582.250000    516.000000      43.000000      1.000000      1.000000   
50%     2300.000000    831.000000      63.000000      2.000000      1.000000   
75%     3500.000000   1300.000000     100.000000      3.000000      2.000000   
max     9990.000000   9990.000000  474002.000000      9.000000      9.000000   

            garagem  
count  20050.000000  
mean       1.212020  
std        1.038961  
min        0.000000  
25%        1.000000  
50%        1.000000  
75%        2.000000  
max        9.000000

We already see something weird with the `area`: the minimum value if `1` squared meter. The only reasonable assumption would be if that was an storage unit.

In [53]:
df[df['area'] == 1]

aluguel  condominio  area                            bairro  quartos  \
4059    2150.0       720.0   1.0                           Cambuci      3.0   
4613    5000.0      2061.0   1.0                      Vila Olímpia      2.0   
5833    2000.0       600.0   1.0                            Jaçanã      3.0   
5890    2300.0       850.0   1.0                      Vila Gustavo      3.0   
7529    1400.0       600.0   1.0                         Vila Nivi      2.0   
7576    3000.0       525.0   1.0              Jardim das Bandeiras      2.0   
9274    2150.0       720.0   1.0                           Cambuci      3.0   
9455    1300.0       513.0   1.0                 Parque Novo Mundo      2.0   
9492    2000.0      1456.0   1.0                      Vila Mariana      2.0   
10129   2500.0       920.0   1.0                        Itaim Bibi      2.0   
10209   5000.0      2061.0   1.0                      Vila Olímpia      2.0   
10211   8000.0      3414.0   1.0                           Paraíso      3.0   
10829     11.0       940.0   1.0                        Itaim Bibi      1.0   
11618   2100.0       761.0   1.0                    Chácara Klabin      1.0   
11865   1000.0       330.0   1.0                     Jardim Modelo      2.0   
14030    950.0         1.0   1.0  Conjunto Habitacional Jova Rural      2.0   
14763   2500.0         1.0   1.0                      Vila Gustavo      2.0   
15476   1200.0       422.0   1.0                 Parque Edu Chaves      2.0   
16094    700.0       393.3   1.0                          Pirituba      1.0   
17578   3500.0       969.1   1.0                   Jardim Paulista      1.0   
18613   1800.0       518.0   1.0                      Vila Mariana      1.0   
18652   1800.0       518.0   1.0                      Vila Mariana      1.0   
18653   1850.0       719.0   1.0            Vila Guarani(Zona Sul)      3.0   
18849   1280.0       745.0   1.0                       Vila Mazzei      2.0   
19164   2300.0       600.0   1.0             Várzea da Barra Funda      2.0   
19179   2900.0       750.0   1.0                          Perdizes      3.0   
19270   1800.0       715.0   1.0            Vila Guarani(Zona Sul)      2.0   
19302   1200.0         1.0   1.0                 Vila Dom Pedro II      1.0   
19618   1000.0       550.0   1.0                Vila Irmãos Arnoni      2.0   
19691   1700.0      1030.0   1.0                       Vila Carrão      1.0   
19716   1700.0      1030.0   1.0                       Vila Carrão      1.0   

       banheiros  garagem                                anunciante  
4059         2.0      2.0                       lello-mooca-locacao  
4613         3.0      2.0                                re-max-400  
5833         2.0      1.0  js-cabral-empreendimento-e-comercio-ltda  
5890         2.0      2.0  js-cabral-empreendimento-e-comercio-ltda  
7529         1.0      1.0  js-cabral-empreendimento-e-comercio-ltda  
7576         2.0      1.0                             rotta-imoveis  
9274         2.0      2.0                       lello-mooca-locacao  
9455         1.0      1.0                            tawata-imoveis  
9492         1.0      1.0                      vila-mariana-locacao  
10129        2.0      0.0                                re-max-400  
10209        3.0      2.0                                re-max-400  
10211        3.0      2.0                                re-max-400  
10829        1.0      1.0                   lello-pinheiros-locacao  
11618        1.0      1.0                             re-max-optima  
11865        1.0      0.0  js-cabral-empreendimento-e-comercio-ltda  
14030        1.0      1.0  js-cabral-empreendimento-e-comercio-ltda  
14763        2.0      1.0  js-cabral-empreendimento-e-comercio-ltda  
15476        2.0      1.0  js-cabral-empreendimento-e-comercio-ltda  
16094        1.0      0.0                   paulo-jose-boltnn-leite  
17578        1.0      1.0       vitoria-flats-negocios-imobiliarios  
18613   

There are a bunch of apartaments with area 1 but they're not storage units (they have quartos > 1, they have bathrooms and so on). Their price are also quite high.
With that in mind, this can only be an error.

In [54]:
100*len(df[df['area'] == 1])/len(df)

0.1546134663341646

Since these represent only `.15%` of the total data its best if we simply drop them.

But first let's check if the second least value for area is reasonable (so that we might have hope of not repeatedly dropping absurd values)

In [55]:
criteria = df['area'] == df[df['area'] != 1]['area'].min()
df[criteria]

aluguel  condominio  area              bairro  quartos  banheiros  \
19842    850.0       393.3   9.0  Jardim Bonfiglioli      1.0        1.0   

       garagem     anunciante  
19842      0.0  miani-imoveis

We see its a very small apartament (3x3 m), but with a reasonable price for that size. It makes sense.

So we'll just drop rows with area == 1:

In [56]:
df = df[df['area'] != 1]
df.describe()

aluguel    condominio           area       quartos     banheiros  \
count  20019.000000  20019.000000   20019.000000  20019.000000  20019.000000   
mean    2737.613667   1093.163115     143.331735      1.946651      1.813477   
std     1820.001028    974.147688    3508.562638      0.952751      1.231183   
min       10.000000      1.000000       9.000000      1.000000      1.000000   
25%     1590.000000    516.000000      43.000000      1.000000      1.000000   
50%     2300.000000    831.000000      63.000000      2.000000      1.000000   
75%     3500.000000   1300.000000     100.000000      3.000000      2.000000   
max     9990.000000   9990.000000  474002.000000      9.000000      9.000000   

            garagem  
count  20019.000000  
mean       1.212148  
std        1.039484  
min        0.000000  
25%        1.000000  
50%        1.000000  
75%        2.000000  
max        9.000000

What about the maximum values for aluguel, condominio and area?

In [57]:
df[df['aluguel'] == df['aluguel'].max()]

aluguel  condominio   area          bairro  quartos  banheiros  \
12647   9990.0      1690.0  320.0  Cidade Monções      3.0        3.0   
15093   9990.0      1690.0  233.0  Cidade Monções      3.0        3.0   
15722   9990.0      1690.0  162.0        Brooklin      3.0        2.0   

       garagem                     anunciante  
12647      2.0                 lullis-imoveis  
15093      2.0  scheid-consultores-de-imoveis  
15722      2.0                     every-home

In [58]:
df[df['condominio'] == df['condominio'].max()]

aluguel  condominio    area               bairro  quartos  banheiros  \
15521     75.0      9990.0  1147.0  Vila Nova Conceição      5.0        5.0   

       garagem                         anunciante  
15521      5.0  imobiliaria-prime-brokers-ltda-me

In [59]:
df[df['area'] == df['area'].max()]

aluguel  condominio      area   bairro  quartos  banheiros  garagem  \
13317     22.0      7400.0  474002.0  Paraíso      4.0        5.0      5.0   

                                              anunciante  
13317  primeiro-imovel-planejamento-e-consultoria-imo...

The first two are very expensive, but for an initial cleaning we'll leave them as they are.

The row with the maximum area makes less sense.

But we'll leave things as they are, since this is an initial cleaning. A cleaning for analysis and ML (which we do in another notebook) would exclude values above/below some multiple of the Z-score.

Now we'll save this dataframe for further analysis.

In [60]:
df.to_csv('apartamentos_clean.csv', index=False)